src/Compilers/CSharp/Test/Emit3/Semantics/PatternMatchingTests4.cs

```diff
@@ -4829,5 +4842,4429 @@ class D : A { }
+                //         if (s is [42 and not 43, 44 and not 45]) { } // 19, 20
+                Diagnostic(ErrorCode.HDN_RedundantPattern, "45").WithLocation(17, 45),
+                // (18,29): warning CS9275: The pattern is redundant.
+                //         if (s is [not 42 or 43, not 44 or 45] and [var x6, var x7]) { } // 21, 22
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "43").WithLocation(18, 29),
+                // (18,43): warning CS9275: The pattern is redundant.
+                //         if (s is [not 42 or 43, not 44 or 45] and [var x6, var x7]) { } // 21, 22
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "45").WithLocation(18, 43),
+                // (20,29): warning CS9275: The pattern is redundant.
+                //         if (o is (not 42 or 43) and var x10) { } // 23
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "43").WithLocation(20, 29));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75506")]
+        public void RedundantPattern_SyntaxHeuristic_Warning()
+        {
+            var source = """
+C s = null;
+_ = s is not null or { Prop1: 42 }; // 1, 2
+_ = s is not null or { Prop1: 44 or 45 }; // 3, 4, 5, 6
+_ = s is not null or { Prop1: >0, Prop2: 0 }; // 7, 8, 9
+
+_ = s is not null or (>0, _); // 10, 11
+_ = s is not null or (>0, 0); // 12, 13, 14
+
+_ = s is not null or Derived; // 15
+_ = s is not null or C { Prop1: 0 }; // 16, 17
+_ = s is not null or C (>0, _) { Prop1: 0 }; // 18, 19, 20
+
+System.Runtime.CompilerServices.ITuple t = null;
+_ = t is not null or (>0, _); // 21, 22
+_ = t is not null or (>0, 0); // 23, 24, 25
+
+_ = s is not null or [{ Prop1: 42 }, ..]; // 26, 27
+_ = s is not null or [_, .. { Prop1: 42 }]; // 28, 29
+
+class C
+{
+    public int Prop1 { get; set; }
+    public int Prop2 { get; set; }
+    public void Deconstruct(out int i, out int j) => throw null;
+
+    public int Length => throw null;
+    public C this[System.Index i] => throw null;
+    public C this[System.Range r] => throw null;
+}
+class Derived : C { }
+""";
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // (2,22): warning CS9275: The pattern is redundant.
+                // _ = s is not null or { Prop1: 42 }; // 1, 2
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "{ Prop1: 42 }").WithLocation(2, 22),
+                // (2,31): warning CS9275: The pattern is redundant.
+                // _ = s is not null or { Prop1: 42 }; // 1, 2
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "42").WithLocation(2, 31),
+                // (3,22): warning CS9275: The pattern is redundant.
+                // _ = s is not null or { Prop1: 44 or 45 }; // 3, 4, 5, 6
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "{ Prop1: 44 or 45 }").WithLocation(3, 22),
```

Comment:
I'm wondering if warnings on a parent pattern and subpatterns at the same time is something we should try to avoid.

Hidden diagnostics seems fine. But warnings may be a little more disruptive, as users may be scrolling through the error list to figure out what the most relevant ones are.